Automattic/newspack-plugin
# feat(content-gate): metering

### All Submissions:






### Changes proposed in this Pull Request:

Implements metering functionality to the content gate.

<img width="286" alt="image" src="https://user-images.githubusercontent.com/820752/235222986-14de4845-78cd-4ef4-8348-dec17d374ca9.png">

If anonymous metering is set, it will occur on the front end using localStorage in order to work with caching. Once signed in, metering is processed in the backend if the content is restricted for that reader.

Along with implementing metering functionality, this also proposes a few changes to the existing code:
 - Switch the gate editor javascript to use `useSelect` and `useDispatch` hooks instead of `withSelect` and `withDispatch` HOCs for a cleaner looking code
 - Change the integration class and namespace from `WC_Memberships` to `Memberships` so its methods can be used without being confused with Woo's original class
 - Use a configuration array for the editor meta to avoid code repetition
 - Added a static method that retrieves the inline gate content so the metering class can use it

### How to test the changes in this Pull Request:

1. Make sure you have RAS enabled and WC Memberships active with configured membership plans
2. Visit the Engagement wizard and click to configure the Content Gate
3. Confirm you see a new "Metering" panel in the editor
4. Toggle Metering on and confirm the parameters set are: 0 anonymous, 0 registered, weekly reset
5. Change the "Available views for registered readers" to 2 and save
6. Visit a restricted post anonymously and confirm the gate is rendered
7. Sign In to an account that is not under the restricted post plan
8. Visit the post and confirm the full content is rendered
9. Visit another post and confirm it's also not restricted
10. Visit a 3rd restricted post and confirm the gate renders
11. Change the gate style to overlay (or inline if it was already set as overlay)
12. Visit the restricted post and confirm the gate renders as expected
13. Edit the gate and change the "Available views for anonymous readers" to 1
14. Visit a restricted post anonymously and confirm it renders
15. Open DevTools and navigate to the localStorage
16. Confirm you see an item named `newspack_metering` with this schema: `{ content: [ 123 ], expiration: 123457890 }`
17. Visit another post and confirm the gate is rendered

### Other information:
